tests/RobotFramework/tests/pkcs11/private_key_storage.robot

```diff
@@ -31,6 +31,69 @@ Use Private Key in SoftHSM2 using tedge-p11-server
 
     Execute Command    tedge reconnect c8y
 
+Select Private key using PKCS#11 URI
+    [Documentation]    Make sure that we can select different keys and tokens using a PKCS#11 URI.
+    ...    The URI can either point to a specific key, or to a specific token where we will attempt to find a key.
+    ...
+    ...    To ensure that correct key is selected and reduce the need to generate and upload different keys and
+    ...    certificates, we'll only be using one key and we'll only import it to the chosen token, keeping other tokens
+    ...    empty.
+
+    # configure tedge
+    Execute Command    tedge config set c8y.url "$(echo ${C8Y_CONFIG.host} | sed 's|https?://||g')"
+    Execute Command    tedge config set mqtt.bridge.built_in true
+    Execute Command    tedge config set device.cryptoki.mode socket
+
+    Execute Command
+    ...    cmd=printf 'TEDGE_DEVICE_CRYPTOKI_MODULE_PATH=%s\nTEDGE_DEVICE_CRYPTOKI_PIN=%s\nTEDGE_DEVICE_CRYPTOKI_URI=%s\n' "/usr/lib/softhsm/libsofthsm2.so" "123456" "" | sudo tee /etc/tedge/plugins/tedge-p11-server.conf
+    Execute Command    sudo -u tedge /usr/bin/init_softhsm.sh --self-signed --device-id "${DEVICE_SN}" --pin 123456
+
+    # Upload the self-signed certificate
+    Execute Command
+    ...    cmd=sudo env C8Y_USER='${C8Y_CONFIG.username}' C8Y_PASSWORD='${C8Y_CONFIG.password}' tedge cert upload c8y
+
+    Execute Command    tedge reconnect c8y
+
+    # expect failure if we try to use a token that doesnt exist
+    Execute Command
+    ...    cmd=printf 'TEDGE_DEVICE_CRYPTOKI_MODULE_PATH=%s\nTEDGE_DEVICE_CRYPTOKI_PIN=%s\nTEDGE_DEVICE_CRYPTOKI_URI=%s\n' "/usr/lib/softhsm/libsofthsm2.so" "123456" "pkcs11:token=asdf" | sudo tee /etc/tedge/plugins/tedge-p11-server.conf
+    Restart Service    tedge-p11-server
+    Run Keyword And Expect Error    *    Execute Command    tedge reconnect c8y
```

Comment:
Use `exp_exit_code=!0` instead of the Robotframework command, otherwise the tedge reconnect c8y will be run multiple times before it fails.

```suggestion
    Execute Command    tedge reconnect c8y    exp_exit_code=!0
```

Reply:
Resolved by 29c496474acb2fb0daf2325fb456e88ffd926985